system/lib/wasmfs/file_table.h

```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "file.h"
+#include <assert.h>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include <wasi/api.h>
+
+namespace wasmfs {
+
+class FileDescription {
+  std::shared_ptr<File> file;
+  __wasi_filedelta_t offset;
+  std::mutex mutex;
+
+public:
+  FileDescription(uint32_t offset, std::shared_ptr<File> file) : offset(offset), file(file) {}
+
+  class Handle {
+    FileDescription& fileDescription;
+    std::unique_lock<std::mutex> lock;
+
+  public:
+    Handle(FileDescription& fileDescription)
+      : fileDescription(fileDescription), lock(fileDescription.mutex) {}
+
+    std::shared_ptr<File>& getFile();
+  };
+
+  Handle get() { return Handle(*this); }
+};
+
+class FileTable {
+  static std::vector<std::shared_ptr<FileDescription>> entries;
+  std::mutex mutex;
+
+  FileTable();
```

Comment:
Why does this have a different pattern of constructor than the previous class?

Reply:
Will have to see about the name change of `FileDescription` too.